File: server/lib/game_machine/handlers/authentication.rb

module GameMachine
  module Handlers
    class Authentication
      AUTHTOKEN = 'authorized'
      AUTHENTICATED_USERS = java.util.concurrent.ConcurrentHashMap.new

      def self.authenticated?(player)
        if authtoken = AUTHENTICATED_USERS.fetch(player.id,nil)
          return player.authtoken == authtoken
        end
        false
      end

      def self.unregister_player(player_id)
        AUTHENTICATED_USERS.delete(player_id)
      end

      def register_player(player)
        AUTHENTICATED_USERS[player.id] =
          authtoken_for_player(player)

      end

      def authtoken_for_player(player)
        if public?
          player.authtoken
        else
          Handlers::PlayerAuthentication.instance.authtoken_for(player.id)
        end
      end

      def public?
        Handlers::PlayerAuthentication.instance.public?
      end

      def valid_authtoken?(player)
        public? || player.authtoken == authtoken_for_player(player)
      end


      def authenticate!(player)
        @player = player
        if valid_authtoken?(player)
          register_player(player)
          player.set_authenticated(true)
        else
          GameMachine.logger.warn("Authentication for #{player.id} failed")
          false
        end
      end

    end

  end
end

File: server/java/server/src/main/java/io/gamemachine/authentication/Authentication.java
package io.gamemachine.authentication;

import io.gamemachine.core.PlayerService;
import io.gamemachine.messages.Player;

import java.util.concurrent.ConcurrentHashMap;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.google.common.base.Strings;

public class Authentication {

    private static ConcurrentHashMap<String, Integer> authenticatedUsers = new ConcurrentHashMap<String, Integer>();
    private static final Logger logger = LoggerFactory.getLogger(Authentication.class);


    public Authentication() {
    }

    public static boolean isAuthenticated(Player player) {
        return (authenticatedUsers.containsKey(player.id));
    }

    public static boolean isAuthenticated(String playerId) {
        return (authenticatedUsers.containsKey(playerId));
    }

    public static boolean hasValidAuthtoken(String playerId, int authtoken) {
        Player player = new Player();
        player.id = playerId;
        player.authtoken = authtoken;
        return Authentication.hasValidAuthtoken(player);
    }

    public static boolean hasValidAuthtoken(Player player) {
        if (authenticatedUsers.containsKey(player.getId())) {
            int authtoken = authenticatedUsers.get(player.id);
            if (authtoken == player.authtoken) {
                return true;
            } else {
                // Reload from database to see if it's been updated
                authtoken = PlayerService.getInstance().getAuthtoken(player.id);
                if (authtoken == player.authtoken) {
                    authenticatedUsers.put(player.id, authtoken);
                    return true;
                }
            }
        }
        return false;
    }

    public static void unregisterPlayer(String playerId) {
        authenticatedUsers.remove(playerId);
    }

    public void registerPlayer(Player player) {
        authenticatedUsers.put(player.getId(), authtokenForPlayer(player));
    }

    public Integer authtokenForPlayer(Player player) {
        if (isPublic()) {
            return player.getAuthtoken();
        } else {
            return PlayerService.getInstance().getAuthtoken(player.id);
        }
    }

    public boolean isPublic() {
        return PlayerAuthentication.getInstance().isPublic();
    }

    public boolean authtokenIsValid(Player player) {
        // 0 signifies logged out in the db, it's reserved and clients should never send this
        if (player.authtoken == 0) {
            return false;
        } else if (player.authtoken == authtokenForPlayer(player)) {
            return true;
        } else if (isPublic()) {
            return true;
        } else {
            return false;
        }
    }

    public boolean authenticate(Player player) {
        if (Strings.isNullOrEmpty(player.id)) {
            return false;
        }
        if (authtokenIsValid(player)) {
            registerPlayer(player);
            player.setAuthenticated(true);
            return true;
        } else {
            logger.warn("Authentication for " + player.getId() + " failed");
            return false;
        }
    }

    public boolean ipCheck(String playerId) {
        Player player = PlayerService.getInstance().find(playerId);
        if (player.ipChangedAt == 0) {  // has not changed since player created
            return true;
        } else if ((System.currentTimeMillis() - player.ipChangedAt) < 60000) {
            return false;
        } else {
            return true;
        }
    }

}
